File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.dart
class Solution {
  String smallestPalindrome(String s) {
    int size = s.length;
    List<String> letters = s.split('');
    Map<String, int> map = Map();
    letters.sort();
    for(int i=0;i<size;i++){
        map.update(letters[i], (val) => val + 1, ifAbsent: () => 1);
    }
    int index = 0;
    if(map.containsValue(1)){
        var t = map.entries.firstWhere(
            (element) => element.value == 1,
        );
        letters[size~/2] = t.key;
        map.remove(t.key);
    }
    map.forEach((key, value) {
        if(value%2 == 1){
            letters[size~/2] = key;
        }
        for(int i=0;i<value~/2;i++){
            letters[index] = key;
            letters[size-1-index] = key;
            index++;
        }
    });
    return letters.join('');
  }
}

File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.cpp
class Solution {
public:
    string smallestPalindrome(string s) {
        int size = s.size();
        string sub = s.substr(0, size/2);
        sort(sub.begin(), sub.end());
        string rever = sub;
        reverse(rever.begin(), rever.end());
        return size % 2 == 1 ? sub + s[size/2] + rever : sub + rever;
    }
};